### PR TITLE
fix(grpc): Allow gRPC connections via Unix socket

### LINKED ISSUE #1832: gRPC server throws ValueError for Unix sockets
**Describe your environment**

```
Ubuntu 20.04
x86_64
Python 3.10.9
```

**Steps to reproduce**

I'll post a PR that addresses the issue with a test that reproduces the problem shortly.
This problem can be triggered by running the `grpc` instrumentation and utilizing Unix sockets for connections.

```py
server.add_insecure_port("unix:///tmp/grpc.sock")
channel = grpc.insecure_channel("unix:///tmp/grpc.sock")
```

**What is the expected behavior?**

The `grpc` instrumentation should collect traces when running on a Unix socket.

**What is the actual behavior?**

A `ValueError` is thrown here:
https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py#L254-L256

`context.peer()` is `unix:` which results in `ip, port = [""]`.

### PR DESCRIPTION
**Pull request by mattoberle**
# Description

The way `NET_PEER_IP` and `NET_PEER_PORT` are retrieved raises a `ValueError` when gRPC connections are handled via Unix sockets.

```py
ip, port = (
    context.peer().split(",")[0].split(":", 1)[1].rsplit(":", 1)
)
```

When using an address like `unix:///tmp/grpc.sock` the value of `context.peer()` is `"unix:"`. Substituting that in the function above...

```py
ip, port = "unix:".split(",")[0].split(":", 1)[1].rsplit(":", 1)
ip, port = ["unix:"][0].split(":", 1)[1].rsplit(":", 1)
ip, port = "unix:".split(":", 1)[1].rsplit(":", 1)
ip, port = ["unix", ""][1].rsplit(":", 1)
ip, port = "".rsplit(":", 1)
ip, port = [""]  # ValueError
```

I addressed the issue by guarding the retrieval of `net.peer.*` values under an `if` statement.
I extended the `server_interceptor` tests to run against TCP and Unix socket configurations.

**Open Questions**

- [ ] The socket tests will fail on Windows. Is there a way to annotate that?
- [ ] Are there other span values we should be setting for the unix socket?

Fixes #1832

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [x] Extended test suite

Removing the `if` statement and running the test suite will reproduce the failure.

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

**Comment by mattoberle** [comment]
Not sure why I keep hitting test failures, they seem unrelated from what I can tell.
Most recent one was for `starlette`.

File: instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py [hunk 1]
```diff
@@ -250,24 +250,30 @@ def _start_span(
...
+                )
 
-            # other telemetry sources add this, so we will too
-            if ip in ("[::1]", "127.0.0.1"):
-                attributes[SpanAttributes.NET_PEER_NAME] = "localhost"
+                # other telemetry sources add this, so we will too
+                if ip in ("[::1]", "127.0.0.1"):
+                    attributes[SpanAttributes.NET_PEER_NAME] = "localhost"
 
-        except IndexError:
```
[the remark this answers is not in the document]
Comment (mattoberle) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:269]: We probably don't want to emit a warning every time we hit this when using unix sockets.
We can move the `if` down into the `except` like tis:

```py
except IndexError, ValueError:
    if context.peer() != "unix:":
        logger.warning(...)
```

That's a little cleaner but comes at a slight cost (doing 3 string extra operations and accessing a few indexes to get there).

I'm open to changing it if you feel strongly either way!

[the remark this answers is not in the document]
Comment (mattoberle) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:269]: Did some very basic analysis.

Comparing timing for the case where `peer == "unix:"`:
```
Using "if" (current in PR)...
0.02746457399916835

Using "try" (proposed in my comment above)...
0.5165281679946929
```

Comparing timing for the case where `peer` is parse-able:
```
Using "if"...
0.17766521099838428

Using "try"...
0.17355161300656619
```

Unix sockets take a noticeable hit when forced to run the extra instructions (there is somewhere around 40 extra instructions to execute). But the more common path (a parse-able peer) is not especially affected in either case.

Granted, this is over 1,000,0000 runs so this optimization is likely negligible.

[the remark this answers is not in the document]
Comment (aabmass) [reply instrumentation/opentelemetry-instrumentation-grpc/src/opentelemetry/instrumentation/grpc/_server.py:269]: I just didn't like the level of nesting but I think your way is more explicit which is nice. Thanks for looking into this

**Review by aabmass — APPROVED** [review]
Thanks for the contribution!

**Review by shalevr — APPROVED** [review]

**Pull request merged**